quinn-proto/src/connection/assembler.rs

```diff
@@ -106,34 +120,55 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
-            let end = offset + (buffer.len() as u64);
-            if let Some(overlap) = end.checked_sub(chunk.offset) {
-                if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
-                    buffer.extend_from_slice(bytes);
+            if chunk.is_fragmented() {
+                let end = offset + (buffer.len() as u64);
+                if let Some(overlap) = end.checked_sub(chunk.offset) {
+                    if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
+                        buffer.extend_from_slice(bytes);
+                    }
+                } else {
+                    self.data.push(Buffer {
+                        offset,
+                        bytes: buffer.split().freeze(),
+                        allocation_size: 0,
+                    });
+                    offset = chunk.offset;
+                    buffer.extend_from_slice(&chunk.bytes);
                 }
             } else {
-                let bytes = buffer.split().freeze();
-                self.data.push(Buffer { offset, bytes });
-                offset = chunk.offset;
-                buffer.extend_from_slice(&chunk.bytes);
+                self.data.push(chunk);
             }
         }
 
         let bytes = buffer.split().freeze();
-        self.data.push(Buffer { offset, bytes });
-        self.defragmented = self.data.len();
-    }
-
-    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes) {
+        self.data.push(Buffer {
+            offset,
+            bytes,
+            allocation_size: 0,
+        });
+        self.allocated = 0;
+        self.buffered = 0
```

Comment:
Shouldn't `buffered` remain unaffected by defragmentation?

Reply:
I've put two comments at the declaration of the `allocated` and `buffered` fields, stating they now only apply to fragmented chunks.

Maybe I should rename the fields to `fragmented_allocated` and `fragmented_buffered`?

Reply:
What seems strange to me here is that the defrag threshold gets smaller after defragmentation. I'd expect the tolerated amount of over-allocation to be proportional to the amount of buffered data in an absolute sense. Perhaps that could be achieved by setting `allocated = buffered` after defrag?

Reply:
OK, I pushed a new version in which `allocated` and `buffered` keep track of both fragmented and defragmented `Buffer`. To prevent defragmenting a `Buffer` twice, I added a `fragmented` field to `Buffer`.